File: backend/app/models/visitor.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base_class import Base
from app.enums.visitor_status import VisitorStatus
from app.enums.visitor_type import VisitorType
from app.enums.entry_mode import EntryMode
from app.enums.approval_mode import ApprovalMode

class Visitor(Base):
    __tablename__ = "visitors"

    id: Mapped[int] = mapped_column(primary_key=True)

    resident_id: Mapped[int] = mapped_column(
        ForeignKey("residents.id"),
        nullable=False,
        index=True,
    )

    visitor_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    phone: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        index=True,
    )

    visitor_type: Mapped[str] = mapped_column(
        String(30),
        default=VisitorType.GUEST.value,
    )

    purpose: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    vehicle_number: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
        index=True,
    )

    expected_time: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(30),
        default=VisitorStatus.PENDING.value,
        index=True,
    )



    # ----------------------------------------------------
    # Walk-in Visitor Information
    # ----------------------------------------------------

    entry_mode: Mapped[str] = mapped_column(
        String(20),
        default=EntryMode.QR.value,
    )

    approval_mode: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
    )

    visitor_photo: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    created_by_guard: Mapped[bool] = mapped_column(
        default=False,
    )

# ----------------------------------------------------
# QR Information
# ----------------------------------------------------
    # ----------------------------------------------------
    # QR Information
    # ----------------------------------------------------

    qr_token: Mapped[str | None] = mapped_column(
        String(100),
        unique=True,
        nullable=True,
    )

    qr_code: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    approved_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    # ----------------------------------------------------
    # Guard Entry Information
    # ----------------------------------------------------

    check_in_time: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
        index=True,
    )

    check_out_time: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

      # ----------------------------------------------------

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        index=True,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )
    
    resident = relationship(
        "Resident",
        back_populates="visitors",
    )

File: frontend/safecolony_app/lib/features/visitors/models/visitor.dart
class Visitor {
  final int id;
  final int residentId;

  final String visitorName;
  final String phone;
  final String visitorType;

  final String? purpose;
  final String? vehicleNumber;

  final String status;

  final String? qrToken;
  final String? qrCode;

  final DateTime? approvedAt;

  final String entryMode;
  final String? approvalMode;
  final String? visitorPhoto;
  final bool createdByGuard;

  Visitor({
    required this.id,
    required this.residentId,
    required this.visitorName,
    required this.phone,
    required this.visitorType,
    this.purpose,
    this.vehicleNumber,
    required this.status,
    this.qrToken,
    this.qrCode,
    this.approvedAt,
    required this.entryMode,
    this.approvalMode,
    this.visitorPhoto,
    required this.createdByGuard,
  });

  factory Visitor.fromJson(Map<String, dynamic> json) {
  return Visitor(
    id: json["id"],
    residentId: json["resident_id"],
    visitorName: json["visitor_name"],
    phone: json["phone"],
    visitorType: json["visitor_type"],
    purpose: json["purpose"],
    vehicleNumber: json["vehicle_number"],
    status: json["status"],

    // New fields
    entryMode: json["entry_mode"] ?? "QR",
    approvalMode: json["approval_mode"],
    visitorPhoto: json["visitor_photo"],
    createdByGuard: json["created_by_guard"] ?? false,

    qrToken: json["qr_token"],
    qrCode: json["qr_code"],
    approvedAt: json["approved_at"] == null
        ? null
        : DateTime.parse(json["approved_at"]),
  );
}

  Map<String, dynamic> toJson() {
    return {
      "resident_id": residentId,
      "visitor_name": visitorName,
      "phone": phone,
      "visitor_type": visitorType,
      "purpose": purpose,
      "vehicle_number": vehicleNumber,
      "entry_mode": entryMode,
      "approval_mode": approvalMode,
      "visitor_photo": visitorPhoto,
      "created_by_guard": createdByGuard,
    };
  }
}